include/awkward/array/NumpyArray.h

```diff
@@ -90,6 +93,8 @@ namespace awkward {
     const NumpyArray getitem_next(const SliceNewAxis& newaxis, const Slice& tail, const Index64& carry, const Index64& advanced, int64_t length, int64_t stride, bool first) const;
     const NumpyArray getitem_next(const SliceArray64& array, const Slice& tail, const Index64& carry, const Index64& advanced, int64_t length, int64_t stride, bool first) const;
 
+    const std::shared_ptr<Content> count(int64_t axis, const NumpyArray* tocarry) const;
```

Comment:
@jpivarski - This is to fold an array into an array, but there must be a better way.

Reply:
I thought I had already implemented `NumpyArray::count`, with and without `axis`. The `count` for a NumpyArray is derived immediately from its `shape`, you don't even need to _look_ at what values it contains, because a NumpyArray is rectilinear. (NumPy itself doesn't have such an operation because it would always be trivial.)

Specifically, for

```python
>>> import numpy
>>> array = numpy.arange(2*3*5).reshape(2, 3, 5)
>>> array
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14]],

       [[15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]]])
```

it would be

```python
>>> def count_numpy(array, axis):
...     return numpy.repeat(array.shape[axis + 1], array.shape[axis])
```

so

```python
>>> count_numpy(array, 0)
array([3, 3])
>>> count_numpy(array, 1)
array([5, 5, 5])
>>> count_numpy(array, 2)
# Error! (As it should be)
```

It's possible to implement something like `numpy.repeat` when the output is a `NumpyArray` by setting the `stride` to `0`:

```python
>>> numpy.lib.stride_tricks.as_strided(numpy.array([5]), shape=(3,), strides=(0,))
array([5, 5, 5])
# Do that in C++ with strides = std::vector<ssize_t>({ 0 })
```

This is saying that the array consists of the same value in memory observed 3 times, rather than actually allocating an array of length 3 and copying the value. That's a nice trick if the output is `NumpyArray`, and therefore `Content`, but not if the output is `Index64`. The indexes are simpler kinds of arrays, used internally by Awkward'd C++, so that we can make stronger assumptions about what they contain and simplify implementations. One of those assumptions is that indexes have stride = itemsize. This can also make a difference in the speed of kernels, since the step size of `for` loops is known at compile-time.

Thus, you can use the stride of zero length trick if you're implementing `std::shared_ptr<Content> count(int64_t axis)`, but not if you're implementing `Index64 count64()`.

---------------------------

About `tocarry`, I don't see any need for carrying in this operation, but if you did have carrying, it should be an `Index64`, not a `NumpyArray`. Indexes have several simplifying features:

   * Indexes are always integers, one of the following: `int8`, `uint8`, `int32`, `uint32`, and `int64`. In some contexts, they're even more limited than that: the `offsets`, `starts`, `stops`, `index`, etc. of ListOffsetArray, ListArray, IndexedArray, etc. can only be `int32`, `uint32`, and `int64`. The `tags` of UnionArray (so far) can only be `int8`. Indexes used for carrying (so far) can only be `int64`.
   * Indexes are always one-dimensional—no shapes or strides.
   * Indexes are always simply contiguous in memory: their "stride" is equal to their itemsize. Since the integer type of an index is a compile-time specialization, the `for` loops over indexes have compile-time step sizes, which has performance advantages.
   * Whereas `NumpyArray` and other `Content` are passed in `std::shared_ptr` to allow runtime virtual function dispatch, indexes are simply passed by reference or copy constructor.

The only complicating features of an index are that they might introduce template specializations (if more than one integer type is allowed, which is only sometimes the case), that they share their numerical data (through a `std::shared_ptr<INT_TYPE>`), and that the meaningful part of the numerical buffer might not start at the beginning (`offset` might be bigger than `0`) or stop at the end (`length` might be less than the allocated buffer's length). This allows slices of an index to not need to be copied—a relevant case for performance—and it complicates calls into kernel functions (the `offsets` and `length` have to be passed separately). If indexes allowed the full complexity of a NumPy array, then _a lot_ more information would have to be passed into every kernel function—this is a balance.

Reply:
Oh, sorry - I'll roll it back and update my branch to the master :-) Thanks for the detailed explanations!